test/FunctionalTests/Infrastructure/DynamicGrpcService.cs

```diff
@@ -0,0 +1,116 @@
+﻿#region Copyright notice and license
+
+// Copyright 2019 The gRPC Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#endregion
+
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using FunctionalTestsWebsite.Infrastructure;
+using Google.Protobuf;
+using Grpc.AspNetCore.Server;
+using Grpc.AspNetCore.Server.Internal;
+using Grpc.Core;
+using Microsoft.AspNetCore.Builder;
+using Microsoft.AspNetCore.Builder.Internal;
+using Microsoft.AspNetCore.Routing;
+using Moq;
+
+namespace Grpc.AspNetCore.FunctionalTests.Infrastructure
+{
+    /// <summary>
+    /// Used by tests to add new service methods.
+    /// </summary>
+    public class DynamicGrpcServiceRegister
+    {
+        private readonly DynamicEndpointDataSource _endpointDataSource;
+        private readonly IServiceProvider _serviceProvider;
+
+        public DynamicGrpcServiceRegister(DynamicEndpointDataSource endpointDataSource, IServiceProvider serviceProvider)
+        {
+            _endpointDataSource = endpointDataSource;
+            _serviceProvider = serviceProvider;
+        }
+
+        public string AddUnaryMethod<TRequest, TResponse, TService>(UnaryServerMethod<TRequest, TResponse> callHandler)
+            where TService : class
+            where TRequest : class, IMessage, new()
+            where TResponse : class, IMessage, new()
+        {
+            var method = CreateMethod<TRequest, TResponse, TService>(MethodType.Unary, Guid.NewGuid().ToString());
```

Comment:
I didn't like having a set of 4 "dynamic" methods that were changed on a test-by-test basis because it runs into trouble when tests execute in parallel.

If test A sets the unary method to Foo, and test B sets the unary method to Bar, and they run at the same time then a race condition occurs.

Also this pattern lets different message types for different tests.